Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp

```diff
@@ -1415,17 +1416,34 @@ void NetworkConnectionToWebProcess::installMockContentFilter(WebCore::MockConten
 #endif
 
 #if ENABLE(LOGD_BLOCKING_IN_WEBCONTENT)
-void NetworkConnectionToWebProcess::logOnBehalfOfWebContent(IPC::DataReference&& logChannel, IPC::DataReference&& logCategory, IPC::DataReference&& logString, uint8_t logType, int32_t pid)
-{
-    OSObjectPtr<os_log_t> osLogChannel;
-    if (logChannel.data() && logCategory.data()) {
-        auto channel = reinterpret_cast<const char*>(logChannel.data());
-        auto category = reinterpret_cast<const char*>(logCategory.data());
-        osLogChannel = adoptOSObject(os_log_create(channel, category));
-    }
-    RELEASE_ASSERT(logType == OS_LOG_TYPE_DEFAULT || logType == OS_LOG_TYPE_INFO || logType == OS_LOG_TYPE_DEBUG || logType == OS_LOG_TYPE_ERROR || logType == OS_LOG_TYPE_FAULT);
+void NetworkConnectionToWebProcess::logOnBehalfOfWebContent(Vector<LogData>&& logs, int32_t pid)
+{
+    static NeverDestroyed<Ref<WorkQueue>> queue(WorkQueue::create("Log Queue", WorkQueue::QOS::Background));
+
+    queue.get()->dispatch([logs = WTFMove(logs), pid = pid, webProcessIdentifier = m_webProcessIdentifier] {
+        for (auto& log : logs) {
+            os_log_message_s msg = { 0 };
+
+            msg.format = log.format.data();
+            msg.buffer = log.buffer.data();
+            msg.buffer_sz = log.buffer.size();
+            msg.privdata = log.privdata.data();
+            msg.privdata_sz = log.privdata.size();
+            
+            char* messageString = os_log_copy_message_string(&msg);
+            if (!messageString)
+                return;
 
-    os_log_with_type(osLogChannel.get() ? osLogChannel.get() : OS_LOG_DEFAULT, static_cast<os_log_type_t>(logType), "WebContent[pid=%d, identifier=%llu]: %s", pid, m_webProcessIdentifier.toUInt64(), logString.data());
+            OSObjectPtr<os_log_t> osLogChannel;
+            if (log.channel.data() && log.category.data())
+                osLogChannel = adoptOSObject(os_log_create(log.channel.data(), log.category.data()));
+            RELEASE_ASSERT(log.type == OS_LOG_TYPE_DEFAULT || log.type == OS_LOG_TYPE_INFO || log.type == OS_LOG_TYPE_DEBUG || log.type == OS_LOG_TYPE_ERROR || log.type == OS_LOG_TYPE_FAULT);
```

Comment:
Shouldn't this be a MESSAGE_CHECK? It's not OK for bad IPC from a compromised Webprocess to kill our shared network process.